docs/string_grader.md

```diff
@@ -1,66 +1,239 @@
+
+When a response doesn't satisfy the given pattern, there are three types of feedback that you can provide, controlled by the `explain_validation` flag:
+
+* The student receives an error message. This does not consume an attempt (`explain_validation='err'`, default)
+* The student is graded incorrect, but receives a message. This does consume an attempt (`explain_validation='msg'`)
+* The student is graded incorrect, and no explanation is given. This does consume an attempt (`explain_validation=None`)
+
+In the first two cases, the message provided is given by the `invalid_msg` flag, which defaults to `Your input is not in the expected format`.
+
+Here is an example of using a validation pattern to accept inputs that look like chemical formulae for organic molecules. Note that anything that matches the pattern will be accepted.
+
+```pycon
+>>> grader = StringGrader(
+...     validation_pattern=r'([CNOH](_[0-9])?)+',
+...     explain_validation='msg',
+...     invalid_msg='Write a chemical formula containing hydrogen, oxygen, carbon and/or nitrogen',
+...     strip_all=True,   # Removes all spaces from the input
+...     accept_any=True
+... )
+>>> grader(None, 'NH_3') == {'grade_decimal': 1,
+...                          'msg': '',
+...                          'ok': True}
+True
+>>> grader(None, 'KCl') == {'grade_decimal': 0,
+...                          'msg': 'Write a chemical formula containing hydrogen, oxygen, carbon and/or nitrogen',
+...                          'ok': False}
+True
 
-```python
-grader = StringGrader(
-    answers={'expect': '', 'grade_decimal': 1, 'msg': 'Your answer has been recorded.'},
-    accept_any=True
-)
 ```
 
-Note that when either of these options are set to True, it is meaningless to provide more than one answer to check against, as they will all be graded as correct. Also note that you do not need to provide any answer at all in this situation.
+Here, we use validation to ensure that the student input matches the desired format before comparing to the answer, and give the student an error message if their input doesn't match the specification.
+
+```pycon
+>>> grader = StringGrader(
+...     answers='(1)(2)',
+...     validation_pattern=r'\([0-9]\)\([0-9]\)',
+...     explain_validation='err',
+...     strip_all=True   # Removes all spaces from the input
+... )
+>>> grader(None, '(1)(2)') == {'grade_decimal': 1, 'msg': '', 'ok': True}
+True
+>>> grader(None, '(2)(1)') == {'grade_decimal': 0, 'msg': '', 'ok': False}
+True
+>>> from mitxgraders.exceptions import InvalidInput
```

Comment:
I moved to `*` imports for the matrix docs so as not to have to import the errors, since importing the error isn't really doc-related.

Reply:
Ok, will do.

Reply:
I had to add `InvalidInput` to `__init__.py` for this to work. Worth it though, I think.